File: 198-house-robber/198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        dp = []

        for idx, i in enumerate(nums):
            if idx == 0:
                dp.append(i)
            elif idx == 1:
                dp.append(max(i, dp[0]))
            else:
                dp.append(max(i+dp[idx-2], dp[idx-1]))

        return dp[len(nums)-1]
        

File: 198-house-robber/198-house-robber.java
class Solution {
    public int rob(int[] nums) {
        //Tabulation dp: O(n)
        int n= nums.length;
        int[] dp= new int[n];
        dp[0]=nums[0];
        for(int i=1; i<n; i++){
            
            int op1= dp[i-1];
            int op2= (i>1) ? dp[i-2]+ nums[i] : nums[i];
            dp[i]= Math.max(op1,op2);
        }
        
        return dp[n-1];        
    }
}